File: solutions/kotlin/prime-factors/1/src/main/kotlin/PrimeFactors.kt
object PrimeFactorCalculator {

    fun primeFactors(int: Int): List<Int> {
        var int = int
        var ans = mutableListOf<Int>()
        var p = 2
        while (int != 1) {
            while (int % p == 0) {
                int /= p
                ans.add(p)
            }
            p += 1
        }
        return ans
    }

    fun primeFactors(long: Long): List<Long> {
        var long = long
        var ans = mutableListOf<Long>()
        var p: Long = 2
        while (long != 1L) {
            while (long % p == 0L) {
                long /= p
                ans.add(p)
            }
            p += 1
        }
        return ans
    }
}

File: solutions/zig/prime-factors/1/prime_factors.zig
const std = @import("std");
const mem = std.mem;

pub fn factors(allocator: mem.Allocator, value: u64) mem.Allocator.Error![]u64 {
    if (value <= 1) {
        return allocator.alloc(u64, 0);
    }
    const prime_factors = try allocator.alloc(u64, 63);
    errdefer allocator.free(prime_factors);
    var v = value;
    var prime: u64 = 2;
    var len: usize = 0;
    for (prime_factors) |*p| {
        if (v == 1) {
            break;
        }

        while (v % prime != 0) {
            prime += 1;
            if (prime * prime > v) {
                prime = v;
            }
        }
        p.* = prime;
        v /= prime;
        len += 1;
    }
    const ans = try allocator.realloc(prime_factors, len);
    return ans;
}
